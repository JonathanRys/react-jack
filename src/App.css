@font-face {
  font-family: "OldStamper";
  src: url("./__fonts__/old_stamper.woff2") format("woff2"),
    url("./__fonts__/old_stamper.woff") format("woff"),
    url("./__fonts__/old_stamper.ttf") format("truetype");
}

.App {
  text-align: center;
}

.App-logo {
  /* animation: App-logo-spin infinite 20s linear; */
  background: url("./__images__/logo.png");
  background-size: cover;
  height: 55px;
  width: 55px;
  display: inline-block;
  margin-bottom: -15px;
  margin-right: 7px;
}

.App-header {
  background-color: #222;
  padding: 20px;
  color: white;
}

.App-title {
  font-size: 1.5em;
  margin-bottom: 0.2em;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
