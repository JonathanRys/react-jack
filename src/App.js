import React, { Component } from "react";

import { Container, Row, Col } from "reactstrap";

import "./App.css";

import PlayField from "./PlayField/PlayField";
import Profile from "./Profile/Profile";
import ControlPanel from "./ControlPanel/ControlPanel";

import { newDeck, shuffle, drawOne, clearCard } from "./actions/deckActions";
import { play, stop, nextPlayer, dealerTurn } from "./actions/turnActions";
import { setName, setAvatar, clearHands } from "./actions/playerActions";
import {
  takeCard,
  dealerTakeCard,
  stand,
  buyChips,
  setBet,
  winBet,
  loseBet,
  setInsured
} from "./actions/playerActions";

import { connect } from "react-redux";

const mapStateToProps = state => {
  return {
    deck: state.deck,
    turn: state.turn,
    player: state.player,
    dealer: state.dealer
  };
};

const dispatchAll = (dispatch, actions) => {
  actions.map(action => {
    return dispatch(action());
  });
};

const mapDispatchToProps = dispatch => {
  return {
    setName: name => {
      dispatch(setName({ name }));
    },
    setAvatar: avatar => {
      dispatch(setAvatar({ avatar }));
    },
    buyChips: newChips => {
      dispatch(buyChips({ newChips }));
    },
    setBet: newBet => {
      dispatch(setBet({ newBet }));
    },
    winBet: multiplier => {
      dispatch(winBet({ multiplier }));
    },
    loseBet: () => {
      dispatch(loseBet());
    },
    play: () => {
      dispatch(play());
    },
    stop: () => {
      dispatch(stop());
    },
    dealerTurn: () => {
      dispatchAll(dispatch, [dealerTurn]);
    },
    nextPlayer: () => {
      dispatch(nextPlayer());
    },
    hitOnClick: () => {
      dispatch(drawOne());
    },
    standOnClick: () => {
      dispatch(stand());
    },
    splitOnClick: () => {
      // Not sure what to do here yet
    },
    doubleDownOnClick: () => {
      // Thunk this
      dispatchAll(dispatch, [drawOne, dealerTurn]);
    },
    buyInsuranceOnClick: () => {
      dispatch(setInsured());
    },
    dealOnClick: () => {
      dispatchAll(dispatch, [clearHands, newDeck, shuffle]);
    },
    keepDealing: () => {
      dispatchAll(dispatch, [nextPlayer, drawOne]);
    },
    giveCard: card => {
      dispatch(takeCard({ card }));
      dispatch(clearCard());
    },
    giveDealerCard: card => {
      dispatch(dealerTakeCard({ card }));
      dispatch(clearCard());
    },
    clearCard: () => {
      dispatch(clearCard());
    },
    clearHands: () => {
      dispatch(clearHands());
    }
  };
};

class App extends Component {
  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1 className="App-title">
            <div className="App-logo" />React Jack
          </h1>
        </header>
        <Container>
          <Row>
            <Col xs="12" xl="3">
              <Profile
                {...this.props.player}
                setName={this.props.setName}
                setAvatar={this.props.setAvatar}
              />
            </Col>
            <Col xs="12" xl="6">
              <PlayField {...this.props} />
            </Col>
            <Col xs="12" xl="3">
              <ControlPanel {...this.props} />
            </Col>
          </Row>
        </Container>
      </div>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(App);
