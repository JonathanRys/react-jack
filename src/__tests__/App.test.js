import React from "react";

import App from "../App";
import { shallow } from "enzyme";

import { createStore } from "redux";
import rootReducer from "../reducers/index";

describe("Test the App component", () => {
  it("matches the snapshot", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);
    expect(component).toMatchSnapshot();
  });

  it("changes the player's name", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().setName("Tester");
    expect(store.getState().player.name).toEqual("Tester");
  });

  it("changes the player's avatar", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().setAvatar("./images/test.png");
    expect(store.getState().player.avatar).toEqual("./images/test.png");
  });

  it("sets the player's status", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().setStatus("Blackjack");
    expect(store.getState().player.status[0]).toEqual("Blackjack");
  });

  it("wins bet", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    expect(store.getState().player.balance).toEqual(500);
    component.props().setBet(5);
    expect(store.getState().player.currentBet).toEqual(5);
    component.props().winBet();
    expect(store.getState().player.balance).toEqual(505);
    component.props().winBet(2);
    expect(store.getState().player.balance).toEqual(515);
  });

  it("deducts given amount from player's balance", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    expect(store.getState().player.balance).toEqual(500);
    component.props().buyChips(500);
    expect(store.getState().player.balance).toEqual(1000);
  });

  it("starts and stops playing the game", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().play();
    expect(store.getState().turn.isPlaying).toEqual(true);

    component.props().stop();
    expect(store.getState().turn.isPlaying).toEqual(false);
  });

  it("sets dealers turn", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().dealerTurn();
    expect(store.getState().turn.playersTurn).toEqual(0);
  });

  it("player hits on click", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    expect(store.getState().deck.drawnCard).toEqual(null);
    component.props().hitOnClick();
    expect(store.getState().deck.drawnCard).not.toEqual(null);
  });

  it("player nextHand on click", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().nextHand();
    expect(store.getState().player.handIndex).toEqual(1);
    component.props().nextHand();
    expect(store.getState().player.handIndex).toEqual(0);
  });

  it("player stands on click", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().standOnClick();
    expect(store.getState().player.playerStands[0]).toEqual(true);
  });

  it("cycles the player and doubles down on click", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    expect(store.getState().player.doubledDown[0]).toEqual(false);
    expect(store.getState().deck.drawnCard).toEqual(null);
    expect(store.getState().turn.playersTurn).toEqual(1);
    component.props().doubleDownOnClick();
    expect(store.getState().player.doubledDown[0]).toEqual(true);
    expect(store.getState().deck.drawnCard).not.toEqual(null);
    expect(store.getState().turn.playersTurn).toEqual(1);
  });

  it("buys insurance on click", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    expect(store.getState().player.hasInsurance[0]).toEqual(false);
    component.props().buyInsuranceOnClick();
    expect(store.getState().player.hasInsurance[0]).toEqual(true);
  });

  it("deals the first card on click", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    // Initial expectations
    expect(store.getState().player.hands[0]).toEqual([]);
    expect(store.getState().deck.deck).toEqual([]);

    store.dispatch({ type: "TAKE_CARD", payload: { card: "H8" } });
    expect(store.getState().player.hands[0]).toEqual(["H8"]);

    store.dispatch({ type: "NEW_DECK" });
    expect(store.getState().deck.deck).not.toEqual([]);

    const unShuffledDeck = store.getState().deck.deck;

    component.props().dealOnClick();
    expect(store.getState().player.hands[0]).toEqual([]);
    expect(store.getState().deck.deck).not.toEqual([]);
    expect(store.getState().deck.deck).not.toEqual(unShuffledDeck);

    component.props().keepDealing();
    expect(store.getState().deck.drawnCard).not.toEqual("");
    expect(store.getState().turn.playersTurn).toEqual(0);
  });

  it("takes player card from draw pile", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().dealOnClick();
    component.props().hitOnClick();
    component.props().nextPlayer();
    const drawnCard = store.getState().deck.drawnCard;

    component.props().giveCard(drawnCard);
    expect(store.getState().player.hands[0]).toEqual([drawnCard]);
    expect(store.getState().deck.drawnCard).toEqual(null);
  });

  it("takes dealer card from draw pile", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().dealOnClick();
    component.props().hitOnClick();
    const drawnCard = store.getState().deck.drawnCard;

    component.props().giveDealerCard(drawnCard);
    expect(store.getState().dealer.hand).toEqual([drawnCard]);
    expect(store.getState().deck.drawnCard).toEqual(null);
  });

  it("clears the hands", () => {
    const store = createStore(rootReducer);
    const component = shallow(<App store={store} />);

    component.props().giveCard("D2");
    component.props().giveDealerCard("H7");

    expect(store.getState().player.hands[0]).toEqual(["D2"]);
    expect(store.getState().dealer.hand).toEqual(["H7"]);

    component.props().clearHands();
    expect(store.getState().player.hands[0]).toEqual([]);
    expect(store.getState().dealer.hand).toEqual([]);
  });
});
