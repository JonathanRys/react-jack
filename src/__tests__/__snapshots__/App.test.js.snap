// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test the App component matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(App)
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "buyChips": [Function],
      "buyInsuranceOnClick": [Function],
      "clearHands": [Function],
      "dealOnClick": [Function],
      "dealer": Object {
        "avatar": "./images/dealer.png",
        "busted": false,
        "hand": Array [],
        "handFlipped": false,
        "hasBlackjack": false,
        "name": "Dealer",
        "score": 0,
      },
      "dealerTurn": [Function],
      "deck": Object {
        "deck": Array [],
        "drawnCard": null,
      },
      "doubleDownOnClick": [Function],
      "giveCard": [Function],
      "giveDealerCard": [Function],
      "hitOnClick": [Function],
      "keepDealing": [Function],
      "nextHand": [Function],
      "nextPlayer": [Function],
      "play": [Function],
      "player": Object {
        "avatar": "./images/default.png",
        "balance": 500,
        "busted": Array [
          false,
        ],
        "currentBet": 5,
        "doubledDown": Array [
          false,
        ],
        "handIndex": 0,
        "hands": Array [
          Array [],
        ],
        "hasBlackjack": Array [
          false,
        ],
        "hasInsurance": Array [
          false,
        ],
        "name": "Player1",
        "playerIndex": 1,
        "playerStands": Array [
          false,
        ],
        "score": Array [
          0,
        ],
        "splitHand": false,
        "status": Array [
          null,
        ],
      },
      "setAvatar": [Function],
      "setBet": [Function],
      "setName": [Function],
      "setStatus": [Function],
      "splitOnClick": [Function],
      "standOnClick": [Function],
      "stop": [Function],
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
      "turn": Object {
        "errMsg": "",
        "error": false,
        "isPlaying": false,
        "numPlayers": 2,
        "playersTurn": 1,
      },
      "winBet": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "buyChips": [Function],
        "buyInsuranceOnClick": [Function],
        "clearHands": [Function],
        "dealOnClick": [Function],
        "dealer": Object {
          "avatar": "./images/dealer.png",
          "busted": false,
          "hand": Array [],
          "handFlipped": false,
          "hasBlackjack": false,
          "name": "Dealer",
          "score": 0,
        },
        "dealerTurn": [Function],
        "deck": Object {
          "deck": Array [],
          "drawnCard": null,
        },
        "doubleDownOnClick": [Function],
        "giveCard": [Function],
        "giveDealerCard": [Function],
        "hitOnClick": [Function],
        "keepDealing": [Function],
        "nextHand": [Function],
        "nextPlayer": [Function],
        "play": [Function],
        "player": Object {
          "avatar": "./images/default.png",
          "balance": 500,
          "busted": Array [
            false,
          ],
          "currentBet": 5,
          "doubledDown": Array [
            false,
          ],
          "handIndex": 0,
          "hands": Array [
            Array [],
          ],
          "hasBlackjack": Array [
            false,
          ],
          "hasInsurance": Array [
            false,
          ],
          "name": "Player1",
          "playerIndex": 1,
          "playerStands": Array [
            false,
          ],
          "score": Array [
            0,
          ],
          "splitHand": false,
          "status": Array [
            null,
          ],
        },
        "setAvatar": [Function],
        "setBet": [Function],
        "setName": [Function],
        "setStatus": [Function],
        "splitOnClick": [Function],
        "standOnClick": [Function],
        "stop": [Function],
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "unsubscribe": [Function],
        },
        "turn": Object {
          "errMsg": "",
          "error": false,
          "isPlaying": false,
          "numPlayers": 2,
          "playersTurn": 1,
        },
        "winBet": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): Object {
    "storeSubscription": undefined,
  },
}
`;
